util/regexp-assemble/regexp-assemble.py

```diff
@@ -9,140 +9,216 @@
-# prefix, suffix, flags, block start block end
-special_comment_markers = '^$+><'
-simple_comment_regex = re.compile(r'^##![^' + special_comment_markers + r'].*')
-
-def detect_preprocessor(line):
-	match = preprocessor_regex.match(line)
-	if match is None:
-		return
-		
-	definition = match.group(1).split()
-	try:
-		processor_type, script_path = preprocessor_map[definition[0]]
-		return processor_type(script_path, definition[1:])
-	except KeyError:
-		print(f'No processor found for {definition}')
-		sys.exit(1)
-
-def _is_simple_comment(line):
-	return simple_comment_regex.match(line) is not None
-
-def preprocess(lines):
-	iterator = lines.__iter__()
-	final_lines = []
-	for processor_type in (LinePreprocessor, BlockPreprocessor, FilePreprocessor):
-		transformed_lines = []
-		line = next(iterator, None)
-		while line is not None:
-			if (_is_simple_comment(line)):
-				line = next(iterator, None)
-				continue
-			processor = detect_preprocessor(line)
-			if processor is None or not isinstance(processor, processor_type):
-				transformed_lines.append(line)
-				line = next(iterator, None)
-				continue
-			transformed_lines += processor.run(iterator)
-			line = next(iterator, None)
-
-		final_lines = transformed_lines
-		iterator = final_lines.__iter__()
-	return final_lines
-
-
-def assemble(lines):
-	args = [os.path.join(lib_directory, 'regexp-assemble.pl')]
-	outs = None
-	errs = None
-	proc = Popen(args, stdin=PIPE, stdout=PIPE, stderr=PIPE)
-	for line in lines:
-		proc.stdin.write(line.encode('utf-8'))
-		proc.stdin.write(b'\n')
-	try:
-		outs, errs = proc.communicate(timeout=30)
-	except TimeoutExpired:
-		proc.kill()
-		print(f'Assembling regex timed out')
-		print('Stderr: ' + errs.decode('utf-8'))
-		sys.exit(1)
-
-	if errs:
-		print('Failed to assemble regex')
-		print('Stderr: ' + errs.decode('utf-8'))
-		sys.exit(1)
-	
-	sys.stdout.write(outs.decode('utf-8'))
-
-def run():
-	if len(sys.argv) > 1 and re.fullmatch(r'\d{6}(?:[^.]+)?', sys.argv[1]):
-		with open(os.path.join(script_directory, 'data', sys.argv[1] + '.data')) as handle:
-			iterator = handle.readlines().__iter__()
-	else:
-		iterator = fileinput.input()
-
-	lines = preprocess(iterator)
-	if len(lines) == 0:
-		print('No input. Either pass a filename, a rule id or pipe data to the script')
-		sys.exit(1)
-	assemble(lines)
-
-if __name__ == '__main__':
-	run()
+import os, re, sys
+import argparse
+import logging
+from typing import TypeVar
+
+from lib.operators.assembler import Assembler
+from lib.operators.comparer import Comparer
+from lib.operators.updater import Updater
+from lib.context import Context
+
+S = TypeVar("S", bound="argparse._SubParserAction[argparse.ArgumentParser]")
+
+
+def build_args_parser() -> argparse.ArgumentParser:
+    parser = argparse.ArgumentParser(
+        description="Utilities for generating and managing regular expressions in CRS rule files"
+    )
+    levels = ("DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL")
+    parser.add_argument(
```

Comment:
What does the `--log-level` do? I couldn't find any information or different behavior when I used.

Reply:
There are some error logs, one info log and a few debug logs. Depending on the configuration you'll see them, default is INFO. It put this mainly because I wanted to have debug output for a specific situation (cmdline escaping).

So yes, to see the difference you currently have to go down a certain code path.

Reply:
> There are some error logs, one info log and a few debug logs.

sorry for the dumb question, but where are these logs? Or is there any other cmd option, that I have to use?

Reply:
```bash
util/regexp-assemble/lib/operators/parser.py
45:        self.logger.info("Processing %s", rule_id)
util/regexp-assemble/lib/operators/assembler.py
116:            self.logger.error("Assembling regex timed out")
121:            self.logger.error("Failed to assemble regex")
122:            self.logger.error("Stderr: %s", errs)
util/regexp-assemble/lib/processors/cmdline.py
54:        self.logger.debug("cmdline in:  %s", line)
55:        self.logger.debug("cmdline out: %s", processed)
```

So with `--loglevel debug` you would see the two outputs form `cmdline.py`, if your data file happens to use the `cmdline` preprocessor.

Reply:
Added `--loglevel` documentation to readme.